### DISCUSSION TITLE
Think about using `orlangure/gnomock` for testing

### DISCUSSION THREAD

**Discussion by muety** [muety/wakapi#729] [Ideas]
[orlangure/gnomock](https://github.com/orlangure/gnomock) looks like a super convenient tool for testing! Test code and external dependencies (e.g. MySQL in our case) could be coupled together as one. Would love to give this a try, but am not yet decided if or how we could actually integrate this.

Our **unit tests** don't require external services, because it's unit tests.

Our [**migration tests**](https://github.com/muety/wakapi/blob/1d5f51dcad86976c0f02022ac79c40d1f32aeec5/testing/run_api_tests.sh#L28) could potentially be implement using gnomock instead, but doesn't really add a lot of benefit.

The [**api tests**](https://github.com/muety/wakapi/tree/master/testing) are probably what I'd most like to implement with gnomock, to get rid of the current Postman / newman setup. But I don't think that's a viable option either, because we'd probably still need a fully-fledged Wakapi server to be running, including entire web- / routing- / middleware stack, etc. 

Any thoughts on this, @YC? :thinking:

**Comment by YC** [comment]
With migration tests, I think there will be value of using something like `gnomock` if the tests are more granular, e.g. seed some data, apply a particular set of migrations and verify the result (and repeat for different DBs). 

From experiences at work, I think that Postman / newman collections don't scale well, so fully agree with moving away from it. 
I think API testing is usually performed with the same instance of service/database throughout a run (poor performance and cannot parallelise otherwise?), so perhaps not a good use of `gnomock`. 
Not familiar with Golang, so can't give recommendations about frameworks to use for API testing. I have some experience with using JS test frameworks to test various things, but that's probably not idiomatic. 

Happy holidays!
